core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,477 @@
+        ConnectivityState newState = stateInfo.getState();
+        // If we have shutdown channels or updated addresses, we may still receive state updates
+        // from subchannels. To prevent pickers from returning these obselete subchannels, this logic
+        // is included to check if the current list of active subchannels includes the current one.
+        if (!subchannels.containsValue(subchannel)) {
+          return;
+        }
+        if (newState == SHUTDOWN) {
+          return;
+        }
+        if (newState == IDLE) {
+          helper.refreshNameResolution(); // TODO: check cases for refresh name resolution
+        }
+
+        // If we are transitioning from a TRANSIENT_FAILURE to CONNECTING or IDLE we ignore this state
+        // transition and still keep the LB in TRANSIENT_FAILURE state. This is referred to as "sticky
+        // transient failure". Only a subchannel state change to READY will get the LB out of
+        // TRANSIENT_FAILURE. If the state is IDLE we additionally request a new connection so that we
+        // keep retrying for a connection.
+
+        // With the new pick first implementation, iterative requests for connections will not be
+        // requested once the first pass through is complete. This means that individual subchannels
+        // are responsible for coming out of backoff and starting a transport.
+        // However, if the first pass through is complete and we report TRANSIENT_FAILURE but an address
+        // update occurs, the iterative logic will still be present for the first pass through for
+        // the new address list even though we are in a state of TRANSIENT_FAILURE.
+        // TODO: this logic has to be implemented, we can just keep a bit that tells us to iterate or not
+        if (currentState == TRANSIENT_FAILURE) {
+          if (newState == CONNECTING || newState == TRANSIENT_FAILURE) {
+            // each subchannel is responsible for its own backoff
+            return;
+          } else if (newState == IDLE) {
+            // coming out of backoff
+            addressIndex.reset();
+            requestConnection();
+            return;
+          }
+        }
+        SubchannelPicker picker;
+        switch (newState) {
+          case IDLE:
+            // Subchannel was shutdown when ready: connect from beginning when prompted
+            addressIndex.reset();
+            picker = new RequestConnectionPicker(subchannels.get(addressIndex.getCurrentAddress()));
+            updateBalancingState(IDLE, picker);
+            break;
+          case CONNECTING:
+            // It's safe to use RequestConnectionPicker here, so when coming from IDLE we could leave
+            // the current picker in-place. But ignoring the potential optimization is simpler.
+            picker = new Picker(PickResult.withNoResult());
+            updateBalancingState(CONNECTING, picker);
```

Comment:
If previously in READY, we should ignore this.
when we are iterating through the list, each of them will report CONNECTING before TF or READY, should we consider optimizing here to not call updateBalancingState if we are already in CONNECTING?

Reply:
1. Do we ever go from READY to CONNECTING? I am thinking that a connection either wouldn't be requested if we are READY which means we wouldn't go into a CONNECTING state, or that we would first have to go into failure/idle state to get to CONNECTING.
2. That makes sense! The only thing I was worried about was `L83` which supplies a picker with a subchannel when it is connecting: 
`updateBalancingState(CONNECTING, new Picker(PickResult.withSubchannel(subchannels.get(addressIndex.getCurrentAddress()))));`
I'm fairly confused as to why this was previously present, and I remember that you pointed it out as well. But if this is the desired behavior, not updating the balancing state/picker would mean that the picker returns the wrong subchannel even when we are looking at a subsequent one later on.